neuralpp/symbolic/z3_expression.py

```diff
@@ -0,0 +1,335 @@
+
+
+def z3_object_to_expression(z3_object: z3.ExprRef) -> Z3Expression:
+    # Here we just try to find a type of expression for z3 object.
+    if z3_object.num_args() > 0:
+        return Z3FunctionApplication(z3_object)
+    elif z3.is_int_value(z3_object) or z3.is_rational_value(z3_object) or z3.is_fp_value(z3_object) or \
+            z3.is_true(z3_object) or z3.is_false(z3_object):
+        return Z3Constant(z3_object)
+    else:
+        return Z3Variable(z3_object)
+
+
+# On the type of *arguments:
+# https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values
+def apply_python_callable_on_z3_arguments(python_callable: Callable,
+                                          *arguments: z3.BoolRef | z3.ArithRef) -> z3.ExprRef:
+    """
+    Directly calling this function can do something that python_callable_to_z3_function(python_callable)(arguments)
+    cannot do:
+        `python_callable_to_z3_function(builtins.min)(x, y)` raises an error, because "min" cannot be turned
+        into a z3 function.
+    while
+        `apply_python_callable_on_z3_arguments(builtins.min, x, y)` is fine, because "min(x,y)" can be turned
+        into a z3 function application (namely, If(x<y, x, y))
+    """
+    match python_callable:
+        # boolean operation
+        case operator.and_:
+            return z3.And(arguments)
+        case operator.or_:
+            return z3.Or(arguments)
+        case operator.not_:
+            return z3.Not(arguments)
+        case operator.xor:
+            return z3.Xor(arguments[0], arguments[1])
+        # comparison
+        case operator.le:
+            return arguments[0] <= arguments[1]
+        case operator.lt:
+            return arguments[0] < arguments[1]
+        case operator.ge:
+            return arguments[0] >= arguments[1]
+        case operator.gt:
+            return arguments[0] > arguments[1]
+        case operator.eq:
+            return arguments[0] == arguments[1]
+        # arithmetic
+        case operator.add:
+            return arguments[0] + arguments[1]
+        case operator.mul:
+            return arguments[0] * arguments[1]
+        case operator.pow:
+            return arguments[0] ** arguments[1]
+        # min/max
+        case builtins.min:
+            return z3.If(arguments[0] < arguments[1], arguments[0], arguments[1])
+        case builtins.max:
+            return z3.If(arguments[0] > arguments[1], arguments[0], arguments[1])
+        case _:
+            raise ValueError(f"Python callable {python_callable} is not recognized.")
+
+
+class Z3Expression(Expression, ABC):
+    def __init__(self, z3_object: z3.ExprRef | z3.FuncDeclRef):
+        super().__init__(get_type_from_z3_object(z3_object))
+        self._z3_object = z3_object
+
+    @classmethod
+    def new_constant(cls, value: Any, type_: Optional[ExpressionType] = None) -> Z3Constant:
+        if isinstance(value, z3.ExprRef | z3.FuncDeclRef):
```

Comment:
Here we touch on the issue of `value` needing to be a Python object everywhere. We probably need all `Z3Expression` implementations (same for `SymPyExpression` implementations) to contains an attribute `base_object` or some such name of type `z3.ExprRef` | `z3.FunDeclRef` (a sympy object type in the case of `SymPyExpression`) and then a property `value` which returns the corresponding Python object.

Note that the same `base_object` can be used to represent the underlying library (Z3 or SymPy) object even for `Z3FunctionApplication` and `Z3Variable`  (`SymPyFunctionApplication` and `SymPyVariable`), but there it is used to produce the appropriate properties (function, arguments, name, etc)